### PR TITLE
[Merge stable to main] Llama3.3-70b and 3.1-8b - Fix sampling parameters

### PR DESCRIPTION
**Pull request by djordje-tt**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/36325

### Problem description
This PR fixes couple of different issues for Llama3.3-70b:
- Non-uniform seeding
- Penalty trap bug
- Penalty bugs for Llama3.1-8b
- batched prefill determinism
- diff between batched and non-batched prefill
- missing logprobs support for Llama3.3-70b

- Fixes same sampling parameters for Llama3.1-8b
### What's changed
- Bring over the log-probs support for Galaxy (optional log-softmaxed logits output), matching the behavior already validated on stable in TT-Metal, vLLM nightly, and Models CI.
- Integrate the deterministic seeding flow (host-side RNG + SamplingSeedManager + `ttnn.manual_seed` usage before `ttnn.sampling`) so prefill + decode produce deterministic sequences across repeats when seeds are fixed.
- Ensure the penalties path matches the shared implementation, fixing the earlier divergence across users.
- Updated matmul configs to support same behaviour across batched and non-batched prefill with couple additional fixes for divergence.

Performance numbers on text_demo in t/s/u: 
| branch | without penalties | with penalties |
|-------|-------|-------|
| branch  | 71.88  t/s/u | 42.36  t/s/u |
| main  | 72.05  t/s/u | - |

**TTFT**:
**68.5**ms -> **73.9**ms drop due to disabling use_2d_grid in rms norm is expected.

### Checklist

- [ ] [All post-commit tests](https://github.com/tenstorrent/tt-metal/actions/runs/21355526046)

#### Model tests

- [x] [Galaxy Demo](https://github.com/tenstorrent/tt-metal/actions/runs/21361481284)
- [x] [vllm nightly](https://github.com/tenstorrent/tt-metal/actions/runs/21361542050)
- [x] [Models CI](https://github.com/tenstorrent/tt-shield/actions/runs/21435406349/job/61728802475)

Last pipelines list 6th Feb: 
-  [] [vllm Nightly](https://github.com/tenstorrent/tt-metal/actions/runs/21754091798)
-  [] [Shield CI](https://github.com/tenstorrent/tt-shield/actions/runs/21753926206/job/62758873631)
-  [] [Galaxy Demo](https://github.com/tenstorrent/tt-metal/actions/runs/21754402409)

### CODE REVIEW

**Comment by yieldthought** [comment]
Codex review:

The patch introduces at least one runtime break (missing SamplingGenerator.reset_seed) and a logic regression in batched prefill logits. It also changes a critical API keyword in log-prob calculation that can corrupt results when log_probs are enabled.

Full review comments:

- [P1] Preserve SamplingGenerator.reset_seed API — [models/common/sampling/generator.py:64](https://github.com/tenstorrent/tt-metal/blob/01544b5cdccfa4d4b99ec26c83bf2057e7102ad4/models/common/sampling/generator.py#L64)
  SamplingGenerator now only initializes a SeedManager, but the public `reset_seed` method was removed; existing call sites such as `models/tt_transformers/tt/generator.py` still invoke `sampling_module.reset_seed(...)` when sampling on device. That call will raise `AttributeError` at runtime for any decode that uses sampling_params. Either keep a wrapper method on SamplingGenerator or update all call sites to use `seed_manager.reset_seed()/get_new_values()`.

- [P2] Avoid batched prefill when returning logits — [models/demos/llama3_70b_galaxy/tt/generator.py:192](https://github.com/tenstorrent/tt-metal/blob/01544b5cdccfa4d4b99ec26c83bf2057e7102ad4/models/demos/llama3_70b_galaxy/tt/generator.py#L192)
  The new `use_batched_prefill` condition no longer checks `return_logits`/`tt_out_logits_all_users`, so batched prefill now runs even when callers request host logits (sampling_params=None). In that path only `tt_out_logits_all_users[id]` (id=0 in batched mode) is populated, leaving other users' logits as zeros. This yields incorrect logits for batch>=16/seq_len=128. Reintroduce the guard or fill logits for every user in batched mode.

- [P2] Use rounding_mode for log-probs chip-id div — [models/common/utils.py:215](https://github.com/tenstorrent/tt-metal/blob/01544b5cdccfa4d4b99ec26c83bf2057e7102ad4/models/common/utils.py#L215)
  In LogProbsCalculator._prepare_relevant_logits, `ttnn.div` is called with `round_mode="floor"`. The TTSampling/TTNN API uses `rounding_mode`; the new keyword is ignored or errors, so chip IDs can be computed with floating division and the remainder/log_probs become incorrect when `enable_log_probs` is true. Use `rounding_mode="floor"` here to keep integer chip IDs.

**Review by dpopovTT — APPROVED** [review]

File: .github/workflows/vllm-nightly-tests-impl.yaml [hunk 1]
```diff
@@ -407,7 +407,7 @@ jobs:
       volumes:
         - ${{ github.workspace }}/docker-job:/work # Subdir to workaround https://github.com/actions/runner/issues/691
         - /dev/hugepages-1G:/dev/hugepages-1G
-        - ${{ (matrix.test-group.runner-label == 'BH-QB-GE' && '/localdev/blackhole_demos/huggingface_data:/mnt/MLPerf/huggingface:ro') || '/mnt/MLPerf:/mnt/MLPerf:ro' }}
+        - ${{ (matrix.test-group.runner-label == 'BH-QB-GE' && '/localdev/blackhole_demos/huggingface_data:/mnt/MLPerf/huggingface:rw') || '/mnt/MLPerf:/mnt/MLPerf:rw' }}
```
Comment (dpopovTT) [inline .github/workflows/vllm-nightly-tests-impl.yaml]: Our best practice here is to set it by default to read only and then have an input in workflow dispatch that changes this to write permissions. We do it like this so we don't change stuff on weka automatically, but on purpose so we are aware of the change

**Review by sraizada-tt — APPROVED** [review]

**Review by rdraskicTT — APPROVED** [review]

File: models/common/tests/test_sampling.py [hunk 2]
```diff
@@ -72,3 +72,94 @@ def test_log_probs_calculation(shape, ttnn_mesh_device):
     print(f"pcc={pcc}")
 
     assert passing, f"Assertion failed, PCC={pcc}"
+
+
+@pytest.mark.parametrize(
+    "shape",
+    [
+        # [1, 1, 256, 256],  # llama on TG and T3K
+        # [1, 1, 32, 8 * 64],  # llama on N300
```
Comment (jvegaTT) [inline models/common/tests/test_sampling.py]: Since this has a hardcoded mesh device for Galaxy, should you delete these two?

File: models/common/tests/test_sampling.py [hunk 3]
```diff
@@ -72,3 +72,94 @@ def test_log_probs_calculation(shape, ttnn_mesh_device):
...
+    tt_log_probs = log_probs_calculator.calculate_log_probs(logits_tensor, ttnn_indices_tensor)
+    log_probs_tt_host = ttnn.to_torch(tt_log_probs, mesh_composer=ttnn.ConcatMeshToTensor(mesh_device, dim=3))
+    # slice from (1,1,32,256) -> (1,1,1,32)
+    log_probs_tt_host = log_probs_tt_host[:, :, :1, :32]
+
+    log_probs_torch = F.log_softmax(torch_tensor.float(), dim=-1)
+    log_probs_torch_argmax = torch.gather(log_probs_torch, dim=-1, index=argmax_tensor)
+    log_probs_torch_argmax = torch.reshape(log_probs_torch_argmax, (1, 1, 1, 32))
+
+    passing, pcc = comp_pcc(log_probs_torch_argmax, log_probs_tt_host, pcc=0.99)
```
Comment (jvegaTT) [inline models/common/tests/test_sampling.py:160]: Why only 2 9s of PCC?

Comment (djordje-tt) [reply models/common/tests/test_sampling.py:160]: Due to lower precision (bfp8b) and only 32 values being compared pcc could go lower than 0.999 for example. We are doing in bfp8b since for higher precision we see larger perf drop with logprobs.

File: models/common/tests/test_sampling.py (hunk 2 above)
Comment (djordje-tt) [reply models/common/tests/test_sampling.py]: Will remove it!

File: models/demos/llama3_70b_galaxy/demo/text_demo.py [hunk 4]
```diff
@@ -1291,7 +1320,7 @@ def test_demo_text(
     }
     # TODO This is suppose to check the config `repeat2`. Since right now that config is the only using a repeat_batches=2 this if statement works
     if repeat_batches == 2 and batch_size == 1:
-        target = 68.50 if galaxy_type == "6U" else 99
+        target = 74.00 if galaxy_type == "6U" else 99
```
Comment (jvegaTT) [inline models/demos/llama3_70b_galaxy/demo/text_demo.py:1323]: Nice improvement!

File: models/demos/llama3_70b_galaxy/tt/llama_ccl.py [hunk 5]
```diff
@@ -526,21 +558,33 @@ def get_ring_prefill_reduce_scatter_buffers(self):
...
-                    "WO": [(1, 1, seqlen, 1280), (1, 1, seqlen, 1280 // 8)],
+                    # "WO": [(1, 1, seqlen, 1280), (1, 1, seqlen, 1280 // 8)],
                     "FF1": [(1, 1, seqlen, 3200), (1, 1, seqlen, 3200 // 4)],
                     "FF3": [(1, 1, seqlen, 3200), (1, 1, seqlen, 3200 // 4)],
                     "FF2": [(1, 1, seqlen, 1280), (1, 1, seqlen, 1280 // 8)],
+                    "QKV_batched": [(1, 32, seqlen // 32, 1280), (1, 32, seqlen // 32, 1280 // 4)],
+                    # "WO_batched": [(1, 32, seqlen // 32, 1280), (1, 32, seqlen // 32, 1280 // 8)],
+                    "FF1_batched": [(1, 32, seqlen // 32, 3200), (1, 32, seqlen // 32, 3200 // 4)],
+                    "FF3_batched": [(1, 32, seqlen // 32, 3200), (1, 32, seqlen // 32, 3200 // 4)],
+                    "FF2_batched": [(1, 32, seqlen // 32, 1280), (1, 32, seqlen // 32, 1280 // 8)],
```
Comment (jvegaTT) [inline models/demos/llama3_70b_galaxy/tt/llama_ccl.py:587]: Consider deleting the W0 and W0_batched references rather than commenting them out? Unless you are going to need those for a subsequent PR

File: models/common/tests/test_sampling.py [hunk 6]
```diff
@@ -12,23 +12,23 @@
...
 @pytest.mark.parametrize(
     "shape",
-    [  # TODO: Add llama3.1-8b T3K shapes
+    [
         [1, 1, 32, 8 * 18992],  # Qwen3 on T3K
     ],
 )
 @pytest.mark.parametrize(
-    "ttnn_mesh_device",
+    "device_params",
```
Comment (gwangTT) [inline models/common/tests/test_sampling.py:20]: `ttnn_mesh_device` is new fixture introduced in TTTv2 that allows device sharing in CI testing. 

Can this test still use `ttnn_mesh_device`? 

Also, this PR should be checked in T3K Unit Tests CI pipeline -- [t3000-unit-tests / t3k_tttv2_fast_unit_tests](https://github.com/tenstorrent/tt-metal/actions/runs/21711108271/job/62615513641#logs), where it runs this: 
https://github.com/tenstorrent/tt-metal/blob/b79a873e21852fe892d459fdedef8eb336d209ea/tests/scripts/t3000/run_t3000_unit_tests.sh#L654

File: models/demos/llama3_70b_galaxy/tt/llama_ccl.py [hunk 7]
```diff
@@ -1230,7 +1286,6 @@ def tt_distributed_rmsnorm(
     compute_kernel_config,
     tt_ccl=None,
 ):
-    # TODO: remove this once we have a proper 2D grid implementation, currently we see bad outputs
```
Comment (jvegaTT) [inline models/demos/llama3_70b_galaxy/tt/llama_ccl.py:1238]: Shouldn't we keep this comment since use_2d_grid is still false?

File: models/tt_transformers/tt/model_config.py [hunk 8]
```diff
@@ -1333,8 +1333,66 @@ def _get_xattn_kv_prefill_mem_cfg(seq_len):
             self.set_tg_attention_config()
 
             self.is_multichip = self.num_devices > 1
+            self.num_reduce_scatter_links = 1
+            self.num_all_gather_links = (
+                2 if self.is_galaxy else 1
+            )  # TODO: try out 3 for short axis and 4 for long axis (TG only) <- should work but untested in model
```
Comment (jvegaTT) [inline models/tt_transformers/tt/model_config.py:1339]: why only one link???

File: models/tt_transformers/tt/model_config.py [hunk 9]
```diff
@@ -1333,8 +1333,66 @@ def _get_xattn_kv_prefill_mem_cfg(seq_len):
...
+                "num_workers_per_link": 2,
+                "rs_memory_config": ttnn.DRAM_MEMORY_CONFIG,
+            }
+            default_sampling_force_argmax = {
+                "allow_force_argmax": False,
+                "num_links": 1,
+                "chunks_per_sync": 10,
+                "num_workers_per_link": 2,
+                "topology": ttnn.Topology.Linear,
+            }
```
Comment (jvegaTT) [inline models/tt_transformers/tt/model_config.py:1357]: Why 1 link with 2 workers?, 2 links with 1 worker would be much much better in nearly every case

**Review by gwangTT — APPROVED** [review]
I left a comment for consideration. Approved to unblock.

File: models/demos/llama3_70b_galaxy/demo/text_demo.py (hunk 4 above)
Comment (djordje-tt) [reply models/demos/llama3_70b_galaxy/demo/text_demo.py:1323]: This is actually TTFT that regressed due to disabling use_2d_grid in rms norm.

File: ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_async/device/kernels/minimal_default_writer.cpp [hunk 10]
```diff
@@ -42,12 +42,16 @@ constexpr uint32_t output_tensor_Ht = get_compile_time_arg_val(14);
...
 constexpr uint32_t reverse = get_compile_time_arg_val(17) == 1;
+#ifdef USE_WORKER_MUX
 constexpr uint8_t fabric_mux_num_buffers_per_channel = get_compile_time_arg_val(18);
 constexpr size_t fabric_mux_channel_buffer_size_bytes = get_compile_time_arg_val(19);
 constexpr size_t fabric_mux_status_address = get_compile_time_arg_val(20);
 constexpr size_t fabric_mux_termination_signal_address = get_compile_time_arg_val(21);
 constexpr uint32_t num_mux_clients = get_compile_time_arg_val(22);
 constexpr uint32_t rt_arg_count = 23;
+#else
+constexpr uint32_t rt_arg_count = 18;
```
Comment (jvegaTT) [inline ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_async/device/kernels/minimal_default_writer.cpp:53]: Should be called ct_arg_count since it is compile time variables it is counting

**Review by jvegaTT — APPROVED** [review]
Approving to unblock but please run blackhole post commit before merging to ensure there are no blackhole related changes to all gather due to this PR.

**Review by mtairum — APPROVED** [review]

File: models/tt_transformers/tt/model.py [hunk 11]
```diff
@@ -612,9 +611,4 @@ def forward(
             x = ttnn.interleaved_to_sharded(x, self.model_config["LM_HEAD_INPUT_MEMCFG"])
 
         x = self.lm_head(x)
-
-        if mode == "prefill":
```
Comment (alingTT) [inline models/tt_transformers/tt/model.py:616]: why is it safe to remove this?

File: models/tt_transformers/tt/model.py [hunk 12]
```diff
@@ -156,7 +156,6 @@ def process_logits_after_prefill_trace(self, logits, last_token_idx):
         if self.model_config["LM_HEAD_INPUT_MEMCFG"].is_sharded():
             logits = ttnn.interleaved_to_sharded(logits, self.model_config["LM_HEAD_INPUT_MEMCFG"])
         logits = self.lm_head(logits)
-        logits = ttnn.to_layout(logits, layout=ttnn.ROW_MAJOR_LAYOUT, memory_config=ttnn.DRAM_MEMORY_CONFIG)
```
Comment (alingTT) [inline models/tt_transformers/tt/model.py:159]: Is it safe to remove this as well? is logits already in DRAM? does downstream usage of this function continue to expect ROW MAJOR? please double check nothing breaks here

Comment (djordje-tt) [reply models/tt_transformers/tt/model.py:159]: Pipelines like Llama3.1-8b demo, vllm nightly for Llama3.1-8b and Models CI for Llama3.1-8b are passing. I can double check but I am pretty confident it's not needed anymore!
